src/platform/cert_capi.c

```diff
@@ -697,6 +697,177 @@ CxPlatCertParseChain(
+    HCERTSTORE TempCertStore = NULL;
+    CERT_BLOB Blob = {0};
+
+    PortableCertificate->Buffer = CertCtx->pbCertEncoded;
+    PortableCertificate->Length = CertCtx->cbCertEncoded;
+    CertificateChain->Buffer = NULL;
+    CertificateChain->Length = 0;
+
+    EnhKeyUsage.cUsageIdentifier = 0;
+    EnhKeyUsage.rgpszUsageIdentifier = NULL;
+    CertUsage.dwType = USAGE_MATCH_TYPE_AND;
+    CertUsage.Usage = EnhKeyUsage;
+    ChainPara.cbSize = sizeof(CERT_CHAIN_PARA);
+    ChainPara.RequestedUsage = CertUsage;
+
+    if (!CertGetCertificateChain(
+            NULL,  // default chain engine
+            CertCtx,
+            NULL,
+            NULL,
+            &ChainPara,
+            0,
+            NULL,
+            &ChainContext)) {
+        LastError = GetLastError();
+        Status = LastError;
+        QuicTraceEvent(
+            LibraryErrorStatus,
+            "[ lib] ERROR, %u, %s.",
+            LastError,
+            "CertGetCertificateChain failed");
+        goto Exit;
+    }
+
+    TempCertStore =
+        CertOpenStore(
+            CERT_STORE_PROV_MEMORY,
+            X509_ASN_ENCODING | PKCS_7_ASN_ENCODING,
+            0,
+            CERT_STORE_ENUM_ARCHIVED_FLAG,
+            NULL);
+    if (NULL == TempCertStore) {
+        LastError = GetLastError();
+        Status = LastError;
+        QuicTraceEvent(
+            LibraryErrorStatus,
+            "[ lib] ERROR, %u, %s.",
+            LastError,
+            "CertOpenStore failed");
+        goto Exit;
+    }
+
+    for (DWORD i = 0; i < ChainContext->cChain; ++i) {
+        PCERT_SIMPLE_CHAIN SimpleChain = ChainContext->rgpChain[i];
+        for (DWORD j = 0; j < SimpleChain->cElement; ++j) {
+            PCERT_CHAIN_ELEMENT Element = SimpleChain->rgpElement[j];
+            PCCERT_CONTEXT EncodedCert = Element->pCertContext;
+            if (!CertAddCertificateLinkToStore(
+                    TempCertStore,
+                    EncodedCert,
+                    CERT_STORE_ADD_ALWAYS,
+                    NULL)) {
+                LastError = GetLastError();
+                Status = LastError;
+                QuicTraceEvent(
+                    LibraryErrorStatus,
+                    "[ lib] ERROR, %u, %s.",
+                    LastError,
+                    "CertAddCertificateLinkToStore failed");
+                goto Exit;
+            }
+        }
+    }
+
+    if (!CertSaveStore(
+            TempCertStore,
+            X509_ASN_ENCODING | PKCS_7_ASN_ENCODING,
+            CERT_STORE_SAVE_AS_PKCS7,
+            CERT_STORE_SAVE_TO_MEMORY,
+            &Blob,
+            0)) {
+        LastError = GetLastError();
+        Status = LastError;
+        QuicTraceEvent(
+            LibraryErrorStatus,
+            "[ lib] ERROR, %u, %s.",
+            LastError,
+            "CertSaveStore failed");
+        goto Exit;
+    }
+
+    Blob.pbData = CXPLAT_ALLOC_NONPAGED(Blob.cbData, QUIC_POOL_TLS_PFX);
```

Comment:
I think we should have a new pool tag here so we can distinguish between the two if a leak occurs.

Reply:
Its not actually possible (At least currently). The PFX code only exists in openssl, and this code only exists in schannel.

Reply:
Ah, right. This is fine then.